### ISSUE TITLE
TC001-3 runtime-evaluated-decorators not working with generic framework decorators

### ISSUE THREAD

**Issue by hartungstenio** [astral-sh/ruff#21282]
### Summary

Pydantic AI's `FunctionToolSet` is a `Generic` class providing a `@toolset.tool` decorator.

`runtime-evaluated-decorators` works without providing the generic type: https://play.ruff.rs/5267d988-584d-4a3e-88d9-4b48cac7fbbc
```python
toolset = FunctionToolset()
```

but it is now working when providing the generic type: https://play.ruff.rs/4e5c8e8c-f76a-4e77-91a5-3c89ca0624d1
```python
toolset = FunctionToolset[Deps]()
```

### Version

0.14.3

**Comment by Daverball** [comment]
This like #20637 looks like another limitation of the very basic type inference in ruff. IIRC there's been another issue where inference failed in a different way. We can probably make the inference slightly more sophisticated than it currently is to at least work around some of these edge cases. This case definitely seems easier to support than the case in the linked issue.

What always works as a workaround in these cases is to add `toolset` with its fully qualified name to `runtime-evaluated-decorators` in addition to `FunctionToolset`. You would need to do this anyway, if you want `@toolset.tool` to work correctly if you access it from a different module.

**Comment by willfrey** [comment]
For what it's worth, this is causing me problems with the `pydantic_ai.Agent` decorators along with `pydantic_ai.RunContext` inside of the decorated function signatures.

```toml
[tool.ruff.lint.flake8-type-checking]
runtime-evaluated-decorators = [
    "pydantic_ai.Agent.system_prompt",
    "pydantic_ai.Agent.output_validator",
    "pydantic_ai.Agent.tool",
    "pydantic_ai.Agent.instructions",
]
```

will work when an agent is created like

```python
agent = Agent("...", deps_type=Deps, output_type=Output)
```

or

```python
agent: Agent[Deps, Output] = agent("...", deps_type=Deps, output_type=Output)
```

but not for this

```python
agent = Agent[Deps, Output]("...", deps_type=Deps, output_type=Output)
```

For the final case that does not work, this causes `RunContext` to get pushed into the `if TYPE_CHECKING: ...` block, which breaks the runtime evaluation that `pydantic_ai` requires.

Right now, `pydantic_ai` doesn't play super nicely with type alias response types (see the `# type: ignore` comment here: https://ai.pydantic.dev/examples/sql-gen/#example-code) and so the last pattern that doesn't work is a nice way to explicitly type annotate the agent without repeating `Agent` twice.
